d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadLoadBalancer.java

```diff
@@ -0,0 +1,312 @@
+    {
+      case NEW_LB_ONLY:
+        _newLb.getClient(request, requestContext, clientCallback);
+        break;
+      case DUAL_READ:
+
+        _newLb.getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>()
+        {
+          @Override
+          public void onError(Throwable e)
+          {
+            _log.error("Double read failure. Unable to read service properties from: " + serviceName, e);
+          }
+
+          @Override
+          public void onSuccess(ServiceProperties result)
+          {
+            String clusterName = result.getClusterName();
+            _newLb.getLoadBalancedClusterProperties(clusterName, new Callback<Pair<ClusterProperties, UriProperties>>()
+            {
+              @Override
+              public void onError(Throwable e)
+              {
+                _log.error("Dual read failure. Unable to read cluster properties from: " + clusterName, e);
+              }
+
+              @Override
+              public void onSuccess(Pair<ClusterProperties, UriProperties> result)
+              {
+                _log.debug("Dual read is successful. Get cluster and uri properties: " + result);
+              }
+            });
+          }
+        });
+        _oldLb.getClient(request, requestContext, clientCallback);
+        break;
+      case OLD_LB_ONLY:
+      default:
+        _oldLb.getClient(request, requestContext, clientCallback);
+    }
+  }
+
+  @Override
+  public void getLoadBalancedServiceProperties(String serviceName, Callback<ServiceProperties> clientCallback)
+  {
+    switch (getDualReadMode(serviceName))
+    {
+      case NEW_LB_ONLY:
+        _newLb.getLoadBalancedServiceProperties(serviceName, clientCallback);
+        break;
+      case DUAL_READ:
+        _newLb.getLoadBalancedServiceProperties(serviceName, Callbacks.empty());
+        _oldLb.getLoadBalancedServiceProperties(serviceName, clientCallback);
+        break;
+      case OLD_LB_ONLY:
+      default:
+        _oldLb.getLoadBalancedServiceProperties(serviceName, clientCallback);
+    }
+  }
+
+  @Override
+  public void getLoadBalancedClusterProperties(String clusterName,
+      Callback<Pair<ClusterProperties, UriProperties>> callback)
+  {
+    switch (getDualReadMode())
+    {
+      case NEW_LB_ONLY:
+        _newLb.getLoadBalancedClusterProperties(clusterName, callback);
+        break;
+      case DUAL_READ:
+        _newLb.getLoadBalancedClusterProperties(clusterName, Callbacks.empty());
+        _oldLb.getLoadBalancedClusterProperties(clusterName, callback);
+        break;
+      case OLD_LB_ONLY:
+      default:
+        _oldLb.getLoadBalancedClusterProperties(clusterName, callback);
+    }
+  }
+
+  @Override
+  public Directory getDirectory()
+  {
+    DualReadModeProvider.DualReadMode dualReadMode = getDualReadMode();
+    if (dualReadMode == DualReadModeProvider.DualReadMode.DUAL_READ
+        || dualReadMode == DualReadModeProvider.DualReadMode.OLD_LB_ONLY)
```

Comment:
looks like this part can be in a small helper function "isNotReadingXdsOnly" since it's repeated quite many times.